Bottom: e06d8cda98db4e12dce26674115aa70458eae4d8
Top:    d4edc2731701b6ca6a94f9087e44dbb9f76ec25c
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-07-18 11:30:22 -0700

Refresh of fork

---

diff --git a/src/preload.c b/src/preload.c
index 7be40f6..56c6c7d 100644
--- a/src/preload.c
+++ b/src/preload.c
@@ -465,11 +465,15 @@ static int connect_fork(int socket, const struct sockaddr *addr, socklen_t addrl
 {
 	int fd, ret;
 	uint32_t msg;
+	long flags;
 
 	fd = fd_getd(socket);
+	flags = real.fcntl(fd, F_GETFD);
+	real.fcntl(fd, F_SETFD, 0);
 	ret = real.connect(fd, addr, addrlen);
-	if (ret)
+	if (ret) {
 		return ret;
+	}
 
 	ret = real.recv(fd, &msg, sizeof msg, MSG_PEEK);
 	if ((ret != sizeof msg) || msg) {
@@ -477,6 +481,7 @@ static int connect_fork(int socket, const struct sockaddr *addr, socklen_t addrl
 		return 0;
 	}
 
+	real.fcntl(fd, F_SETFD, flags);
 	ret = transpose_socket(socket, fd_rsocket);
 	if (ret < 0)
 		return ret;
